### PR TITLE
Validate boot in JeOS images

### PR DESCRIPTION
**Pull request by mloviska**
- Related ticket: [[qac][JeOS] extend jeos testing to validate UEFI boot with secure boot](https://progress.opensuse.org/issues/64445)
- [osd needles - MOK](https://gitlab.suse.de/openqa/os-autoinst-needles-sles/-/merge_requests/1393)
- [o3 needles - MOK manager](https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/672)
- Verification runs:
   * sle15sp3
      * [w/ sb](http://kepler.suse.cz/tests/1617)
      * [w/o sb](http://kepler.suse.cz/tests/1606)
   * sle15sp2
      * [w/ sb@svirt-hyperv-uefi](http://kepler.suse.cz/tests/839#step/verify_efi_mok/1)
      * [w/o sb@svirt-hyperv-uefi](http://kepler.suse.cz/tests/830#step/verify_efi_mok/1)
      * [w/ sb](http://kepler.suse.cz/tests/837#step/verify_efi_mok/1)
      * [w/o sb](http://kepler.suse.cz/tests/835#step/verify_efi_mok/1)
   * sle15sp1
      * [w/ sb](http://kepler.suse.cz/tests/1618)
      * [w/o sb](http://kepler.suse.cz/tests/1607)
   * tumbleweed
     * [w/ sb](http://kepler.suse.cz/tests/906#step/verify_efi_mok/1)
     * [w/o sb](http://kepler.suse.cz/tests/904#step/verify_efi_mok/1)
   * leap15.2
     * [w/ sb](http://kepler.suse.cz/tests/902#step/verify_efi_mok/1)
     * [w/o sb](http://kepler.suse.cz/tests/903#step/verify_efi_mok/1)
   * leap15.1
     * [w/ sb](http://kepler.suse.cz/tests/1593#step/verify_efi_mok/1)
     * [w/o sb](http://kepler.suse.cz/tests/1594#step/verify_efi_mok/1)

### CODE REVIEW

**Comment by mloviska** [comment]
> I don't quite get what the main purpose(s) of this PR is. If it's just for verifying whether secure boot is enabled, a single command would be sufficient. I don't currently see how the parted parsing code would be useful, is it meant to be some more in-depth verification?

Basically, I am trying to verify default configuration of our EFI setup. The `gpt` part has been already done in [console/gpt_ptable.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/gpt_ptable.pm), although it could possibly fit here as the test retrieves ESP for further usage anyway. Additionally, it checks the alignment of the partition as well. 

I assume that the image should not boot from other efi record than `sles/opensuse-secureboot`. As long as it was not changed intentionally to use other EFI boot record. Thus, I am trying to 'refresh' boot options using pbl reinit or trying to reinstall shim and rebooting the SUT without any bootloader config changes (except of gfxpayload in grub2.cfg). And as it is in sle15sp2/tw, SUT boots in efi mode defaultly but after shim-install it starts to boot from `/EFI/sles/shim.efi ( sles-secureboot )` and pbl reinit does not change it.

Yup, the test is quite unclear and messy, since I am still not confident which would be the proper approach over here. Grepping config files, does not seem to make much sense.

Generally, it would be nice to have test module dedicated to validate booting of our images. So we can put there also other boot related test functions like `verify_mount` in the future.

File: lib/jeos.pm [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+our @EXPORT = qw(expect_mount_by_uuid set_grub_gfxmode rebooting is_efi_boot);
+
+sub expect_mount_by_uuid {
+    return (is_hyperv || is_sle('>=15-sp2') || is_tumbleweed || is_leap('>=15.2'));
+}
+
+# non empty */sys/firmware/efi/* must exist in UEFI mode
+sub is_efi_boot {
+    return !script_run('test -d /sys/firmware/efi/') && script_run('test -z "$(ls -A /sys/firmware/efi/)"');
```
Comment (asmorodskyi) [inline lib/jeos.pm:35]: this would save us from few extra `wait_serial` calls : 
```suggestion
    return !script_run('test -d /sys/firmware/efi/ && test -z "$(ls -A /sys/firmware/efi/)"');
```

Comment (asmorodskyi) [reply lib/jeos.pm:35]: after playing a little with this in terminal I have more to say : 
```suggestion
    return script_run('test -z "$(ls -A /sys/firmware/efi/)"');
```
1. not sure why we need first check at all , second one fully cover expectations ( determine if folder exists and non-empty at the same time ). Do you bother about error message in case folder not exists ? Because it is `script_run` this will not break the test.
2. I confused with inversion `!` here . function name is `is_efi_boot` so I would expect that it returning true when `efi` folder exists and none empty , right ? But `test` returns `1` when expression is true so unless I overlook something we would get 0 when folder exists and not empty with current implementation

Comment (mloviska) [reply lib/jeos.pm:35]: we would need to invert `test -z` to `test -n`, because the directory must be non-empty.
And I am afraid we wouldn't be able to determine that the problem is missing directory or the emptiness of the directory

Comment (mloviska) [reply lib/jeos.pm:35]: 1. https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/9997#discussion_r408014736 and it is in the comment above :)

2. test returns 0 when expression is true, and we need to invert it when in comes to perl

```
An exit status of zero indicates that the expression evaluated as true
       and an exit status of 1 indicates that the expression evaluated as false.
```

Comment (mloviska) [reply lib/jeos.pm:35]: And it there is not existing directory, no need to execute `ls`.

Comment (mloviska) [reply lib/jeos.pm:35]: FYI, this how fedora does it. :)
https://openqa.fedoraproject.org/tests/579214/modules/uefi_postinstall/steps/1/src

Comment (mloviska) [reply lib/jeos.pm:35]:
```
➜  other test -z "$(ls -A /sys/firmware/efi/)"; echo $?
ls: cannot access '/sys/firmware/efi/': No such file or directory
0
➜  other test -z "$(ls -A ./a)"; echo $?
ls: cannot access './a': No such file or directory
0
➜  other mkdir a
➜  other test -z "$(ls -A ./a)"; echo $?
0
➜  other mkdir a/b                      
➜  other test -z "$(ls -A ./a)"; echo $?
1
➜  other test -n "$(ls -A ./a)"; echo $?
0
```

Comment (mloviska) [reply lib/jeos.pm:35]: yup, the *error* message is not nice there

File: lib/jeos.pm [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+sub rebooting {
+    my ($self, $msg) = @_;
+    power_action('reboot', textmode => 1);
+    record_info('reboot', $msg);
+    wait_serial '/^BdsDxe:\s+failed.*Access\sDenied/', expect_not_found => 1 or
```
Comment (asmorodskyi) [inline lib/jeos.pm:42]: can we grep some log file for this instead using wait_serial ? because this would actually means that we freezing our test here for 90 seconds to make sure that all is fine .

File: lib/jeos.pm (hunk 1 above)
Comment (asmorodskyi) [reply lib/jeos.pm:35]: >     2. test returns 0 when expression is true, and we need to invert it when in comes to perl
> 
> 
> ```
> An exit status of zero indicates that the expression evaluated as true
>        and an exit status of 1 indicates that the expression evaluated as false.
> ```

from where you quoting this ? Just run this on my laptop : 
```
13:38-e113$[0] test -z "$(ls -A /home/asmorodskyi/)"
15:32-e113$[0] echo $?
1
```

Comment (asmorodskyi) [reply lib/jeos.pm:35]: after posting this I realized that you , right . I just confuse what is expected here , sorry

File: tests/jeos/check_efi_setup.pm [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    # helper function to retrieve location of efi executable
+    my $get_efi_exec = sub {
+        my $str = shift;
+        $str or die "Empty record in efibootmgr parser!\n";
+        $str =~ s/\\/\//g;
+        return substr $str, index($str, '(') + 1, (index($str, ')') - 1) - index($str, '(');
+    };
+
+    my @ebm_raw = split /\n/, script_output 'efibootmgr --verbose';
+    die "EFI variables are not supported\n" if ($ebm_raw[0] =~ /EFI variables are not supported/);
```
Comment (asmorodskyi) [inline tests/jeos/check_efi_setup.pm:48]: isn't `efibootmgr` will return non-zero error code so `script_output` would die even before this line ?

Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:48]: Good point, I had a feeling that it was not reporting rcs somewhere. Maybe mistaking it with some other util. let me check

File: tests/jeos/check_efi_setup.pm [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    # get current boot no., the last element of the first line
+    my $h->{bootid} = (split /\s+/, shift @ebm_raw)[-1];
+    ($h->{bootid}) or die "No id of current boot found!\n";
+
+    # booted efi details
+    my @boot_entry_tokens = split /\s/, (grep /^boot$h->{bootid}\*.*$/i, @ebm_raw)[0];
+    # remove Boot{$boot_no}
+    shift @boot_entry_tokens;
```
Comment (asmorodskyi) [inline tests/jeos/check_efi_setup.pm:57]: maybe it make sense to log what we are removing , just in case format will change at some point so we will have chance to notice this 
```suggestion
    recod_info('removing Boot{$boot_no} from efibootmgr --verbose output' . shift @boot_entry_tokens);
```

or less verbose version : 
```
    diag('removing Boot{$boot_no} from efibootmgr --verbose output' . shift @boot_entry_tokens);

File: tests/jeos/check_efi_setup.pm (hunk 3 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:48]: Yes, completely right
https://github.com/rhboot/efibootmgr/blob/master/src/efibootmgr.c#L1764

File: tests/jeos/check_efi_setup.pm [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    $self->rebooting('gfxmode set');
+    check_efi_state $efi_p_mp;
+    check_mok;
+    ## shim install
+    assert_script_run 'shim-install';
+    $self->rebooting('After shim-install');
+    check_efi_state $efi_p_mp;
+    check_mok;
+    # grub2 install
+    #assert_script_run "grub2-install --efi-directory=$efi_p_mp --target=x86_64-efi $esp_details->{drive}";
```
Comment (asmorodskyi) [inline tests/jeos/check_efi_setup.pm:207]: I know that PR has WIP label so it is ok . But I would leave a comment as a reminder not to forget to drop this before commit :)

File: tests/jeos/check_efi_setup.pm (hunk 4 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:55]: Mainly we do not have serial terminal on each backend. It would be harder to get this information from the screenshots. In the erial_terminal.txt it would be visible if you would compare few lines.

Was not sure how to approach this properly, I am glad that *diag* is sufficient for others as well.

File: lib/jeos.pm (hunk 2 above)
Comment (mloviska) [reply lib/jeos.pm:42]:
```
localhost login: YpuUt-0-
[   37.284607] reboot: Restarting system
[2J[01;01H[=3h[2J[01;01H[0m[35m[40m[0m[37m[40mBdsDxe: loading Boot0003 "opensuse" from HD(2,GPT,9DD7709A-086D-4B2B-A3B5-3B4451F0D520,0x1800,0x10800)/\EFI\opensuse\grubx64.efi
BdsDxe: failed to load Boot0003 "opensuse" from HD(2,GPT,9DD7709A-086D-4B2B-A3B5-3B4451F0D520,0x1800,0x10800)/\EFI\opensuse\grubx64.efi: Access Denied
BdsDxe: loading Boot0001 "UEFI Misc Device" from PciRoot(0x0)/Pci(0x8,0x0)
BdsDxe: starting Boot0001 "UEFI Misc Device" from PciRoot(0x0)/Pci(0x8,0x0)
UEFI SHIM
$Version: 14 $
$BuildMachine: GNU/Linux $
$Commit: 02e2fc61bd2fb7f0045f15db105de7b8ace3029f $
[0m[37m[44m[01;01H������������������������������������������������������������������������������Ŀ[02;01H�                            Verification succeeded   
```

I haven't found this anywhere. It is on serial line at very early boot stage before grub2 loads. ( I think )

File: tests/jeos/check_efi_setup.pm (hunk 5 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:207]: Yes, we cannot forget about this. I have to clarify this part. According to [yast](https://github.com/yast/yast-bootloader/blob/master/src/lib/bootloader/grub_install.rb#L81), this is the way how to setup efi without SB. I experimented with it a bit, but it failed everywhere. But maybe I have just missed some parameter, if so we can fix. Otherwise I plan to drop it. Or we can replace it by modification of `/etc/sysconfig/bootloader` where we can set `SECURE_BOOT="no"` and use pbl instead of `grub2-install`. Personally I would prefer use generic grub2 tools instead of `pbl`.

**Comment by Vogtinator** [comment]
It is expected that Leap images currently fail after `update-bootloader --reinit`, that's the nature of bsc#1167746. I'd remove the softfail as continuing the test at that point wouldn't lead to a proper boot and it'll be fixed with the next kiwi update (hopefully soon) anyway.

File: tests/jeos/check_efi_setup.pm (hunk 5 above)
Comment (Vogtinator) [reply tests/jeos/check_efi_setup.pm:207]: Are you sure you tested this with secure boot disabled? The command looks good to me.

**Comment by Vogtinator** [comment]
I wonder why this check is only run for JeOS - it seems like it would make sense on non-JeOS as well.

File: tests/jeos/check_efi_setup.pm (hunk 5 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:207]: > Are you sure you tested this with secure boot disabled? The command looks good to me.

Yes, I haven't modified `/etc/sysconfig/bootloader`. I thought that `grub2-mkconfig` would not touch it. I will update the code. Thanks!

Comment (Vogtinator) [reply tests/jeos/check_efi_setup.pm:207]: grub2-mkconfig neither reads nor modifies it. I'm asking about the VM settings. If secure boot is enabled (no matter which keys) `grub2-install` will not result in a bootable system.

Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:207]: DeepinScreenshot_select-area_20200420094317
Got mislead by the label. We would need to change the VM secure boot settings from the test depending on the backend.

File: variables.md [hunk 6]
```diff
@@ -32,6 +32,7 @@ DESKTOP | string | | Indicates expected DM, e.g. `gnome`, `kde`, `textmode`, `xf
 DEPENDENCY_RESOLVER_FLAG| boolean | false      | Control whether the resolve_dependecy_issues will be scheduled or not before certain modules which need it.
 DEV_IMAGE | boolean | false | This setting is used to set veriables properly when SDK or Development-Tools are required.
 DISABLE_ONLINE_REPOS | boolean | false | Enables `installation/disable_online_repos` test module, relevant for openSUSE only. Test module explicitly disables online repos not to be used during installation.
+DISABLE_SECUREBOOT | boolean | faslse | Disable secureboot during runtime of a test job. Used in `jeos/check_uefi.pm`
```
Comment (Vogtinator) [inline variables.md:35]: `faslse`

File: lib/jeos.pm [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      record_info('UEFI', 'Access denied reported by MOK', result => 'fail');
+    $self->wait_boot;
+    $self->select_serial_terminal;
+    ensure_serialdev_permissions;
+}
+
+# Set GRUB_GFXMODE to 1024x768
+sub set_grub_gfxmode {
```
Comment (Vogtinator) [inline lib/jeos.pm:50]: Not really related to this PR, but I wonder why we need this specifically for JeOS.

File: lib/jeos.pm [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+use bootloader_setup qw(change_grub_config grep_grub_settings grub_mkconfig set_framebuffer_resolution set_extrabootparams_grub_conf);
+
+our @EXPORT = qw(expect_mount_by_uuid set_grub_gfxmode rebooting is_efi_boot);
+
+sub expect_mount_by_uuid {
+    return (is_hyperv || is_sle('>=15-sp2') || is_tumbleweed || is_leap('>=15.2'));
+}
+
+# non empty */sys/firmware/efi/* must exist in UEFI mode
+sub is_efi_boot {
```
Comment (Vogtinator) [inline lib/jeos.pm:34]: Not JeOS specific, might fit better somewhere else.

File: lib/opensusebasetest.pm [hunk 9]
```diff
@@ -799,6 +803,7 @@ sub handle_grub {
         send_key 'ctrl-x';
     }
     else {
+        $self->tianocore_secureboot_invert_state if (get_var 'DISABLE_SECUREBOOT');
```
Comment (Vogtinator) [inline lib/opensusebasetest.pm:806]: So if secure boot is for some reason already disabled, this would enable it?

Maybe change to a `tianocore_disable_secureboot` function instead which has a needle match on the disabled setting.

File: lib/jeos.pm [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    return !script_run('test -d /sys/firmware/efi/') && script_run('test -z "$(ls -A /sys/firmware/efi/)"');
+}
+
+sub rebooting {
```
Comment (Vogtinator) [inline lib/jeos.pm:38]: The name is too generic

**Comment by mloviska** [comment]
> I still don't see how this would be specific to JeOS, at least `DISABLE_SECUREBOOT` would also be useful to other test suites.

Generally speaking, I have no intention to hide anything, and I am very happy if people find it useful for their testing. 
But I haven't actively promoted it anywhere beyond JeOS, leaving it open for others to notice by themselves if there is any interest.
Nevertheless, where would you like to store the test case module then ?

File: lib/jeos.pm (hunk 7 above)
Comment (mloviska) [reply lib/jeos.pm:50]: I assume grub resolution does not bother them. 
https://openqa.suse.de/tests/4146546#step/boot_to_desktop/1 and they set video at the beginning of each install job.

File: tests/jeos/check_efi_setup.pm [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+sub run {
+    my $self = shift;
+    $self->select_serial_terminal;
+
+    my $esp_details = get_esp_from_parted;
+    my $pkgs        = 'efivar mokutil sbsigntools';
+    $pkgs .= ' dosfstools' if (is_leap('<15.2') || is_sle('<15-sp2'));
+    zypper_call 'ar --no-gpgcheck https://download.opensuse.org/repositories/home:/jejb1:/UEFI/openSUSE_Tumbleweed/ sb_signtools';
```
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:203]: Not happy about it either ,but it was used in [slenkis test suite](https://gitlab.suse.de/SLEnkins/tests-shim/-/blob/master/testsuite-sut/01_test_sut.sh#L47) and it is mentioned in https://en.opensuse.org/openSUSE:UEFI_Secure_boot_using_qemu-kvm

[the remark this answers is not in the document]
Comment (Vogtinator) [reply tests/jeos/check_efi_setup.pm:203]: Sigh. I assume you could submit it to oS:F (packaging should be easy).
Or for now just remove the use of it...

[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:203]: > Sigh. I assume you could submit it to oS:F (packaging should be easy).
> Or for now just remove the use of it...

Sure, why not?! It seems to be useful.

File: lib/jeos.pm (hunk 7 above)
Comment (mloviska) [reply lib/jeos.pm:45]: The image won't change resolution after it leaves grub2 without updated gfx.
Tested locally and basically that is the same as we do in grub2 when the at the image's first boot. 

https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/eb635250aacc75e880f7fd7e22db518cb12a6a93

File: lib/opensusebasetest.pm (hunk 9 above)
Comment (mloviska) [reply lib/opensusebasetest.pm:806]: That was the initial name I have used during development. But as we are going to use the test module in 2 test suites (sb on, sb off). I will change it back.

The needle located [X] with surrounding black background, but ignored the state inside []. So if the keystroke wouldn't be missed or ignored it should revert the actual state. 

[tianocore-devicemanager-sb-conf-attempt-sb-20200421.png](https://gitlab.suse.de/openqa/os-autoinst-needles-sles/-/blob/master/tianocore-devicemanager-sb-conf-attempt-sb-20200421.png)

File: lib/utils.pm [hunk 12]
```diff
@@ -1682,4 +1683,10 @@ sub assert_file_content {
     my ($param, $value) = @_;
     assert_script_run("cat $param | grep $value");
 }
+
+# non empty */sys/firmware/efi/* must exist in UEFI mode
+sub is_efi_boot {
+    return !script_run('test -d /sys/firmware/efi/') && script_run('test -z "$(ls -A /sys/firmware/efi/)"');
```
[the remark this answers is not in the document]
Comment (mloviska) [reply lib/utils.pm:1689]: I assume that we can trust `test`'s output. Do you have any situation in mind that expression can yield `undef`?

[the remark this answers is not in the document]
Comment (mloviska) [reply lib/utils.pm:1689]: Ok, possible timeout, fair enough! thanks

File: lib/opensusebasetest.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (mloviska) [reply lib/opensusebasetest.pm:806]: I use inside `$self->wait_grub;`, which comes from `opensusebasetest`

File: tests/jeos/check_efi_setup.pm [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    # get current boot no., the last element of the first line
+    my $h->{bootid} = (split /\s+/, shift @ebm_raw)[-1];
+    ($h->{bootid}) or die "No id of current boot found!\n";
+
+    # booted efi details
+    my @boot_entry_tokens = split /\s/, (grep /^boot$h->{bootid}\*.*$/i, @ebm_raw)[0];
+    diag("Removing current boot entry \( $h->{bootid} \) from efibootmgr --verbose output" . shift @boot_entry_tokens);
+
+    # get efi executable only
+    $h->{exec} = $get_efi_exec->((split /\//, pop @boot_entry_tokens)[-1]);
```
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:55]: HD(1,GPT,f69f2504-72e1-4ed5-84f8-9cd7bbe04c4e,0x800,0xfa000)/File(\EFI\opensuse\shim.efi) this gets split by '/' into 2 items, and I want the last one. Then sub get_efi_exec retrieves 'foo' from ('foo').

File: lib/opensusebasetest.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (cfconrad) [reply lib/opensusebasetest.pm:806]: ah ok, I oversee this. I would recommend then rename `$self` to `$basetest`, as `$self` is somehow connected with the class.
and still call it with
```suggestion
        tianocore_secureboot_invert_state($self) if (get_var 'DISABLE_SECUREBOOT');
```

[the remark this answers is not in the document]
Comment (mloviska) [reply lib/opensusebasetest.pm]: That would require change over here as well
https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/opensusebasetest.pm#L789
Is this worth it ?

File: tests/jeos/check_efi_setup.pm [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+    return $h;
+}
+
+sub check_efi_state {
+    is_efi_boot or die "Image did not boot in UEFI mode!\n";
+    my $expected = shift;
+
+    # check UEFI firmware bitness
+    validate_script_output "cat ${ \SYSFS_EFI_BITS }", sub { $_ == 64 };
```
Comment (cfconrad) [inline tests/jeos/check_efi_setup.pm:79]: Maybe this, because of constants
```suggestion
    validate_script_output('cat ' . SYSFS_EFI_BITS, sub { $_ == 64 };
```

File: tests/jeos/check_efi_setup.pm [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        }
+    }
+
+}
+
+sub check_mok {
+    my $state = shift() ? qr/^SecureBoot\senabled$/ : qr/^SecureBoot\sdisabled$/;
+    # check SecureBoot according to MOK
+    diag('Expected regex used to verify SecureBoot: ' . $state);
+    validate_script_output 'mokutil --sb-state', sub { $state };
```
Comment (cfconrad) [inline tests/jeos/check_efi_setup.pm:131]: I think this should work without sub:
```suggestion
    validate_script_output 'mokutil --sb-state', $state;
```
see: https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L1208

File: tests/jeos/check_efi_setup.pm [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+
+    $_ = script_output 'mokutil --list-new', proceed_on_failure => 1;
```
Comment (cfconrad) [inline tests/jeos/check_efi_setup.pm:133]: I would not use these globals if not needed.

File: tests/jeos/check_efi_setup.pm [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+            assert_script_run "grep SECURE_BOOT=\"$state\" ${ \SYSCONFIG_BOOTLADER }";
+            assert_script_run 'update-bootloader --reinit';
+        }
+    );
+}
+
+sub post_fail_hook {
+    select_console('root-console');
+
+    upload_logs("cat ${ \GRUB_DEFAULT }",        'etc_default_grub.txt');
```
Comment (cfconrad) [inline tests/jeos/check_efi_setup.pm:265]: also here:
```suggestion
    upload_logs('cat ' . GRUB_DEFAULT,        'etc_default_grub.txt');
```

File: tests/jeos/check_efi_setup.pm [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+    my $blk_dev_driver = {
+        qemu  => 'virtblk',
+        svirt => {
+            xen    => 'xvd',
+            hyperv => 'scsi'
+        }
+    };
+
+    return ($backend eq 'qemu') ? $blk_dev_driver->{$backend} :
+      $blk_dev_driver->{$backend}->{get_required_var('VIRSH_VMM_FAMILY')};
```
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:156]:
```
my $a = 'qemu';
my $b = '';
my $backend        =  join('_', $a,$b);
my $blk_dev_driver = {
    qemu  => 'virtblk',
    svirt_xen => 'xvd',
    svirt_hyperv => 'scsi'
};
```

Yields `Join test --- qemu_ --- `

[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:156]: The same when `$b = undef`

File: tests/jeos/check_efi_setup.pm [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+# return a slice of parted's output
+sub extract_from_parted {
+    my $p_rec = shift;
+
+    $p_rec =~ s/,\s?/_/g;
+    local $/ = ';';
+    chomp $p_rec;
+    return (split /:/, $p_rec)[@_];
```
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm:166]: Very nice catch! Initially, my intention was to return a slice, thus `@_` contained list of indexes of parted record's split fields which I was interested in. Not needed anymore, we need only the first field ( `[0]` )

File: tests/jeos/check_efi_setup.pm (hunk 18 above)
[the remark this answers is not in the document]
Comment (cfconrad) [reply tests/jeos/check_efi_setup.pm]: Hm true, maybe this then:
```
my $a = 'qemu';
my $b;
my $backend        =  join('_', grep {$_} ($a,$b));
my $blk_dev_driver = {
    qemu  => 'virtblk',
    svirt_xen => 'xvd',
    svirt_hyperv => 'scsi'
};

File: lib/utils.pm [hunk 20]
```diff
@@ -1629,4 +1630,10 @@ sub assert_file_content {
     my ($param, $value) = @_;
     assert_script_run("cat $param | grep $value");
 }
+
+# non empty */sys/firmware/efi/* must exist in UEFI mode
+sub is_efi_boot {
+    return !!script_output('test -d /sys/firmware/efi/ && ls -A /sys/firmware/efi/', proceed_on_failure => 1);
```
[the remark this answers is not in the document]
Comment (mloviska) [reply lib/utils.pm]: Sure, but the goal is to check whether there is any output produced by `ls -A /sys/firmware/efi/`. Since the first part takes care of the error message in case of missing directory, the other one just prints out the content of the directory.

File: tests/jeos/check_efi_setup.pm (hunk 14 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm]: Usual dispute between concatenation vs interpolation in our PRs. :D
I am fine with both, personally I feel we (as test contributors) prefer to use interpolation over concatenation. Does concatenation fit you more ?

File: tests/jeos/check_efi_setup.pm (hunk 15 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm]: Nice catch, thanks!

File: tests/jeos/check_efi_setup.pm (hunk 16 above)
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm]: True, it is not the clean way.

File: tests/jeos/check_efi_setup.pm [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+sub get_esp_from_parted {
+    my @parted_data = split /\n/, script_output 'parted --list --machine --script';
+    # locate drive record_info
+    # old dos partition table is not supported in JeOS
+    my $drive_raw_record = (grep(/gpt/ && qr/virtualized_block_type()/, @parted_data))[0];
+    ($drive_raw_record) or die "No gpt table found!\n";
```
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm]: Personally, I prefer the scalar context over list. What is the motivation to store `virtualized_block_type()`'s value in variable?

File: tests/jeos/check_efi_setup.pm (hunk 14 above)
Comment (cfconrad) [reply tests/jeos/check_efi_setup.pm]: I agree, looks like tabs vs spaces :). And I wouldn't say anything if it would look normal like `"cat $SYSFS_EFI_BITS"` or `"cat ${SYSFS_EFI_BITS}"`. But duo to `constant` we need this `${ \SYSFS_EFI_BITS }` and only this looks like concatenation fits more here. But I don't have hard feelings. Just do as you want.

File: tests/jeos/check_efi_setup.pm (hunk 21 above)
[the remark this answers is not in the document]
Comment (cfconrad) [reply tests/jeos/check_efi_setup.pm]: Cause I'm not sure when perl evaluates perl code in regex and when not. Actually out of my brain I thought you need a special modifier like `e` and now I made some research and I think that `e` modifier is only for substitution `s///`.

From here https://perldoc.perl.org/perlop.html#Regexp-Quote-Like-Operators I would say it should be a STRING.
And then I'm wonder why it is working for you. 

Try this little snipped
```
perl -we "$(cat <<'EOT'
use Mojo::Base -strict;

sub reg { return "FOO"; }
my $x = "BLa FOO blub";
my $r_as_string = reg();

say "#"x80;
say '/reg()/ match' if ($x =~ /reg()/);
say 'qr/reg()/ match' if ($x =~ qr/reg()/);
say 'qr(reg()) match' if ($x =~ qr(reg()));
say 'qr/$r_as_string/ match' if ($x =~ qr/$r_as_string/);
say '=~ reg() match' if ($x =~ reg());
EOT
)"

[the remark this answers is not in the document]
Comment (cfconrad) [reply tests/jeos/check_efi_setup.pm]: It doesn't let me go. Take a look to the output of this snippet:

```
perl -we "$(cat <<'EOT'
use Mojo::Base -strict;
use Data::Dumper;

my @arr = ('aa', 'bb', 'cc', 'ccaa', 'bbaa');
sub r{ return 'cc'; }

say 'Output of: grep (/aa/ && qr/r()/, @arr)';
say(Dumper(grep (/aa/ && qr/r()/, @arr)));

say 'Output of: grep (/aa/ && qr/cc/, @arr)';
say(Dumper(grep (/aa/ && qr/cc/, @arr)));

say 'Output of: grep (/aa/ && /r()/, @arr)';
say(Dumper(grep (/aa/ && /r()/, @arr)));

say 'Output of: grep (/aa/ && /cc/, @arr)';
say(Dumper(grep (/aa/ && /cc/, @arr)));
EOT
)"

This actually makes me thinking that `qr()` doesn't take the global `$_` as matching argument ...
So for your case the `&& qr/virtualized_block_type()/` didn't had any effect, or?

[the remark this answers is not in the document]
Comment (mloviska) [reply tests/jeos/check_efi_setup.pm]: > 
> This actually makes me thinking that `qr()` doesn't take the global `$_` as matching argument ...
> So for your case the `&& qr/virtualized_block_type()/` didn't had any effect, or?

You are completely right! Looks like the `qr` part in `grep` will just be simply always true as any non-empty string :/. Cannot recall how I have tested it though... 2 lessons learned today :D thanks a lot!

File: lib/utils.pm (hunk 20 above)
[the remark this answers is not in the document]
Comment (cfconrad) [reply lib/utils.pm:1653]: Ok, Then I was just confused by the comment `non empty */sys/firmware/efi/* must exist in UEFI mode`

File: tests/console/verify_efi_mok.pm [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+    $settings->{label} = $efi_rec_label;
+
+    return $settings;
+}
+sub efibootmgr_current_boot {
+    my $ebm_raw = script_output 'efibootmgr --verbose';
+    my $h       = {};
+
+    ($h->{bootid}) = $ebm_raw =~ /^BootCurrent:\s+(\d+)/m;
+    die("Missing BootCurrent: in efibootmgr output") unless (exists($h->{bootid}) or defined($h->{bootid}));
```
Comment (cfconrad) [inline tests/console/verify_efi_mok.pm:46]:
```suggestion
    die("Missing BootCurrent: in efibootmgr output") unless (defined($h->{bootid}));
```
as `$h->{bootid}` will exists always but undefined if it hadn't a match.

see:
perl -we "$(cat <<'EOT'
my $h = {};
my $s = "FOO BAR";
($h->{test}) = $s =~ /WILL_NOT(.*)_MATCH/m;
print "exists$/" if (exists($h->{test}));
print "defined$/" if (defined($h->{test}));
EOT
)"

Comment (mloviska) [reply tests/console/verify_efi_mok.pm]: Seems like I should be concerned about `label` key only. Thanks! :)

File: tests/console/verify_efi_mok.pm [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+    my $ebm_raw = script_output 'efibootmgr --verbose';
+    my $h       = {};
+
+    ($h->{bootid}) = $ebm_raw =~ /^BootCurrent:\s+(\d+)/m;
+    die("Missing BootCurrent: in efibootmgr output") unless (defined($h->{bootid}));
+
+    if ($ebm_raw =~ /^(Boot$h->{bootid}\*\s+(.+)\s+(\S+\([^\)]+\)\/?)+.*|Boot$h->{bootid}\*\s+(.+))$/m) {
+        $h->{label} = $2 // $4;
+    }
+    die("Missing lable in Boot$h->{bootid} entry") unless (exists($h->{label}) or defined($h->{label}));
```
Comment (cfconrad) [inline tests/console/verify_efi_mok.pm:51]: same here

Comment (cfconrad) [reply tests/console/verify_efi_mok.pm:51]: ah, my mistake. In this case you can check exists :)

Comment (mloviska) [reply tests/console/verify_efi_mok.pm:51]: Not sure, if the condition fails, I think key `label` won't even exist. But I have typo there :D

**Comment by mloviska** [comment]
> LGTM
> nothing to add from my side. Maybe you could run some newer VR's

I am running them all the time locally, since there are still some changes. I will add them and afterwards change label from `wip` to `ready`.

Thanks for valuable feedback!
